## **Bab 1: Pengenalan dan Pembuatan Array NumPy**

*1.1. Apa itu NumPy dan Keunggulannya?*

**NumPy** (singkatan dari Numerical Python) adalah library fundamental di Python untuk komputasi saintifik. Objek utamanya adalah `ndarray` (n-dimensional array), yaitu sebuah tabel data yang semua isinya harus memiliki **tipe data yang sama**.

Bayangkan sebuah List di Python bisa diisi berbagai macam data (angka, tulisan, dll). Nah, NumPy Array lebih spesifik, seperti sebuah kotak yang hanya bisa diisi oleh satu tipe data saja. Hal ini memberikan beberapa keuntungan besar dibanding List biasa:

- **Kecepatan**: Operasi matematika di NumPy jauh lebih cepat karena dieksekusi menggunakan bahasa C yang lebih efisien.

- **Memori**: NumPy Array membutuhkan ruang memori yang lebih sedikit daripada List Python untuk jumlah data yang sama.

- **Kenyamanan**: Memungkinkan kita melakukan operasi matematika pada seluruh data sekaligus tanpa perlu menulis perulangan (loop). Konsep ini disebut Vectorization.

*1.2. Membuat Array 1 Dimensi dari List*

In [ ]:
# Pertama, kita import library NumPy dengan alias 'np' (ini adalah standar umum)
import numpy as np

# Sebuah list Python biasa
list_python = [1, 2, 3, 4, 5]

# Mengubah list menjadi array NumPy
array_numpy1 = np.array(list_python)

print(f"Ini adalah List Python: {list_python}")
print(f"Ini adalah Array NumPy 1D: {array_numpy1}")
print(f"Tipe data list_python: {type(list_python)}")
print(f"Tipe data array_numpy: {type(array_numpy1)}")

# Perbedaan list biasa dengan ndarray
print(f"Operasi matematika pada list: {list_python * 2}")
print(f"Operasi matematika pada ndarray: {array_numpy1 * 2}")

Ini adalah List Python: [1, 2, 3, 4, 5]
Ini adalah Array NumPy 1D: [1 2 3 4 5]
Tipe data list_python: <class 'list'>
Tipe data array_numpy: <class 'numpy.ndarray'>
Operasi matematika pada list: [1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
Operasi matematika pada ndarray: [ 2  4  6  8 10]


*1.3. Membuat Array 2D dan 3D*

In [ ]:
# Membuat array 2 dimensi (punya baris dan kolom)
array_numpy2 = np.array([[1, 2, 3], [4, 5, 6]])
print(f"Ini adalah Array NumPy 2D:\n {array_numpy2}")
print("\n")

# Membuat array 3 dimensi (dengan bentuk (depth, row, column) = (2, 2, 3))
array_numpy3 = np.array([
    [[1, 2, 3], [4, 5, 6]],
    [[7, 8, 9], [10, 11, 12]]
    ])
print(f"Ini adalah Array NumPy 3D:\n {array_numpy3}")

Ini adalah Array NumPy 2D:
 [[1 2 3]
 [4 5 6]]


Ini adalah Array NumPy 3D:
 [[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]


*1.4. Membuat Array dengan Fungsi Bawaan*

In [ ]:
import numpy as np

# Membuat array berisi angka 0
array_nol = np.zeros(5)
print(f"Array berisi angka nol: {array_nol}")

# Membuat array berisi angka 1
array_satu2 = np.ones((3, 2))
print(f"Array berisi angka satu 2D:\n {array_satu2}")

# Membuat array berisi 1 angka custom yang sama (contoh: 5)
array_custom = np.full((2, 3), 5)
print(f"Array berisi angka custom:\n {array_custom}")

# Membuat matrix identitas
array_eye = np.eye(4)
print(f"Matrix identitas 4x4:\n {array_eye}")

# Membuat array dengan rentang nilai (dari 0 hingga sebelum 10, dengan langkah 2)
# Mirip seperti fungsi range() di Python
array_rentang = np.arange(0, 10, 2)
print(f"Array dengan rentang nilai: {array_rentang}")

# Membuat array berisi angka-angka dengan jarak yang sama besar di antara nilai awal dan nilai akhir (start, end, num)
array_linspace = np.linspace(0, 2, 5)
print(f"Array dengan jarak yang sama: {array_linspace}")

Array berisi angka nol: [0. 0. 0. 0. 0.]
Array berisi angka satu 2D:
 [[1. 1.]
 [1. 1.]
 [1. 1.]]
Array berisi angka custom:
 [[5 5 5]
 [5 5 5]]
Matrix identitas 4x4:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
Array dengan rentang nilai: [0 2 4 6 8]
Array dengan jarak yang sama: [0.  0.5 1.  1.5 2. ]


## **Bab 2: Atribut dan Manipulasi Bentuk Array**

*2.1. Mengetahui Atribut Array*

In [ ]:
import numpy as np

arr_2d = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])

# .ndim: Mengetahui jumlah dimensi
print(f"Jumlah dimensi: {arr_2d.ndim}")

# .shape: Mengetahui bentuk array (jumlah baris, jumlah kolom)
print(f"Bentuk array: {arr_2d.shape}")

# .size: Mengetahui total elemen dalam array
print(f"Total elemen: {arr_2d.size}")

# .dtype: Mengetahui tipe data elemen di dalam array
print(f"Tipe data elemen: {arr_2d.dtype}")

Jumlah dimensi: 2
Bentuk array: (2, 4)
Total elemen: 8
Tipe data elemen: int64


*2.2. Mengubah Bentuk Array (.reshape)*

In [ ]:
import numpy as np

# Array 1D dengan 12 elemen (0 sampai 11)
arr = np.arange(12)
print(f"Array asli (1D): {arr}")

# Mengubah bentuknya menjadi matriks 3 baris dan 4 kolom
arr_reshaped = arr.reshape(3, 4)
print("Array setelah di-reshape (3, 4):\n", arr_reshaped)

Array asli (1D): [ 0  1  2  3  4  5  6  7  8  9 10 11]
Array setelah di-reshape (3, 4):
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


## **Bab 3: Operasi Matematika (Vectorization)**

Vectorization adalah teknik untuk menghindari loop eksplisit (for/while loop) dengan menggantinya menggunakan operasi array yang dijalankan secara paralel di level bawah (biasanya ditulis dalam C di balik layar).

Tujuan Vectorization:
* Lebih cepat (karena pakai optimasi low-level).
* Kode lebih singkat dan lebih mudah dibaca.
* Lebih efisien secara memori.

### *3.1. Operasi Aritmatika Dasar (element-wise)*

In [ ]:
import numpy as np

a = np.array([10, 20, 30, 40])
b = np.array([2, 4, 6, 8])

# Penjumlahan
print("Penjumlahan:", np.add(a, b))

# Pengurangan
print("Pengurangan:", np.subtract(a, b))

# Perkalian
print("Perkalian:", np.multiply(a, b))

# Pembagian
print("Pembagian:", np.divide(a, b))

# Setiap elemen dipangkatkan
print("Power:", np.power(a, 2))

# Potong nilai agar dalam rentang [4,5]
print("Clip:", np.clip(b, 4, 7))

Penjumlahan: [12 24 36 48]
Pengurangan: [ 8 16 24 32]
Perkalian: [ 20  80 180 320]
Pembagian: [5. 5. 5. 5.]
Power: [ 100  400  900 1600]
Clip: [4 4 6 7]


### *3.2. Fungsi Matematika Universal (UFuncs)*

In [ ]:
import numpy as np

a = np.array([-1.0, 0.0, 1.0, 2.0])
b = np.array([2.0, 4.0, 6.0, 8.0])

# Mengembalikan nilai absolut
print(f"Nilai absolut: {np.abs(a)}")

# Menghitung akar kuadrat (square root) setiap elemen
print(f"Akar kuadrat: {np.sqrt(b)}")

# Menghitung eksponensial (e^x) setiap elemen
print(f"Eksponensial: {np.exp(b)}")

# Menghitung logaritma natural setiap elemen
print(f"Logaritma: {np.log(b)}")


Nilai absolut: [1. 0. 1. 2.]
Akar kuadrat: [1.41421356 2.         2.44948974 2.82842712]
Eksponensial: [   7.3890561    54.59815003  403.42879349 2980.95798704]
Logaritma: [0.69314718 1.38629436 1.79175947 2.07944154]


### *3.3. Statistik & Agregasi (menghitung ringkasan statistik)*

In [ ]:
import numpy as np

# Data awal

data_nilai = np.array([50, 85, 92, 65, 78, 95])
data_jam_belajar = np.array([3, 8, 9, 4, 6, 10])

In [ ]:
# Statistik dasar
print(f"Rata-rata nilai: {np.mean(data_nilai):.1f}")
print(f"Median: {np.median(data_nilai)}")
print(f"Median (via percentile): {np.percentile(data_nilai, 50)}")
print(f"Jumlah nilai: {np.sum(data_nilai)}")
print(f"Nilai maksimum: {np.max(data_nilai)}")
print(f"Nilai minimum: {np.min(data_nilai)}")
print(f"Standar deviasi: {np.std(data_nilai):.2f}")
print(f"Varians: {np.var(data_nilai):.2f}")

Rata-rata nilai: 77.5
Median: 81.5
Median (via percentile): 81.5
Jumlah nilai: 465
Nilai maksimum: 95
Nilai minimum: 50
Standar deviasi: 15.73
Varians: 247.58


In [ ]:
# Gabungan data untuk analisis hubungan
data_gabungan = np.vstack((data_nilai, data_jam_belajar)).T
print("Data Gabungan \n(Nilai vs Jam Belajar):\n", data_gabungan)

Data Gabungan 
(Nilai vs Jam Belajar):
 [[50  3]
 [85  8]
 [92  9]
 [65  4]
 [78  6]
 [95 10]]


In [ ]:
# Korelasi & Kovarians
print("Matriks Korelasi:")
print(np.corrcoef(data_gabungan.T))

print("\nMatriks Kovarians:")
print(np.cov(data_jam_belajar, data_nilai))

Matriks Korelasi:
[[1.       0.976327]
 [0.976327 1.      ]]

Matriks Kovarians:
[[  7.86666667  47.2       ]
 [ 47.2        297.1       ]]


In [ ]:
# Index nilai maksimum
print("Index nilai maksimum (argmax):", np.argmax(data_nilai))

Index nilai maksimum (argmax): 5


### *3.4. Logika dan Seleksi*

In [ ]:
import numpy as np

# Menampilkan data unik
# Contoh data preferensi pelanggan:
# 0 = Hitam, 1 = Susu, 2 = Mocha
jenis_kopi = np.array([0, 1, 1, 0, 2, 1])
print("Jenis kopi yang pernah dipilih pelanggan:", np.unique(jenis_kopi))

Jenis kopi yang pernah dipilih pelanggan: [0 1 2]


In [ ]:
# Prediksi probabilitas untuk 2 pelanggan
# Kolom: [Kopi Hitam, Kopi Susu, Kopi Mocha]
data = np.array([
         [0.1, 0.7, 0.2],   # Pelanggan 1
         [0.2, 0.3, 0.5],   # Pelanggan 2
         ])

print("Rekomendasi kopi (rasa dengan peluang tertinggi):", np.argmax(data, axis=1))

Rekomendasi kopi (rasa dengan peluang tertinggi): [1 2]


In [ ]:
# rasa mana saja yang punya kemungkinan tinggi (di atas 50%), sehingga bisa ditampilkan sebagai "Top Picks"
# Jika peluang > 0.5, beri 1 (ditampilkan), sisanya 0 (tidak direkomendasikan)
top_picks = np.where(data > 0.5, 1, 0)
print("Top Picks (peluang > 50%):\n", top_picks)

Top Picks (peluang > 50%):
 [[0 1 0]
 [0 0 0]]


### *3.5. Operasi Aljabar Linier (Linear Algebra)*

Digunakan untuk operasi matematis dasar dalam model ML seperti neural networks, regresi linear, atau PCA.

In [ ]:
import numpy as np

In [ ]:
# Dot product antara dua vektor atau matriks
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])

dot_product = np.dot(a, b)
print("dot product dari 2 vektor:", dot_product)
# output -> (1x4 + 2x5 + 3x6)

dot product dari 2 vektor: 32


In [ ]:
# Matrix multiplication (2D atau lebih)
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

matmul_result = np.matmul(A, B)
print(" Hasil perkalian dengan np.matmul:\n", matmul_result)

# atau pakai operator @
print("Hasil perkalian dengan A @ B:\n", A @ B)

 Hasil perkalian dengan np.matmul:
 [[19 22]
 [43 50]]
Hasil perkalian dengan A @ B:
 [[19 22]
 [43 50]]


In [ ]:
# Invers matriks (hanya untuk matriks bujur sangkar & tidak singular)
M = np.array([[1, 2], [3, 4]])
inv_M = np.linalg.inv(M)
print("Invers matriks:\n", inv_M)

Invers matriks:
 [[-2.   1. ]
 [ 1.5 -0.5]]


In [ ]:
# Menghitung Panjang / Frobenius norm dari vektor atau matriks
v = np.array([3, 4])

norm_v = np.linalg.norm(v)
print("Norma vektor:", norm_v)

Norma vektor: 5.0


### *3.6. Operasi Matriks*

In [ ]:
#Operasi Matriks
import numpy as np
X = np.array([[1, 2, 3],
              [0, 1, 4],
              [5, 6, 0]])
Y = np.array([[1,2,3],
              [4,5,6],
              [7,8,9]])

print("addition")
Z_add=X+Y
print(Z_add)
print("substraction")
Z_subs=X-Y
print(Z_subs)
print("Multiplication")
Z_multi=X*Y
print(Z_multi)
print("dot product")
Z_dot_1=X@Y
print(Z_dot_1)
print("dot product")
Z_dot_2=np.dot(X,Y)
print(Z_dot_2)
print("Matriks Transpose")
Z_transpose=X.T
print(Z_transpose)
print("Matriks Inverse")
Z_inverse=np.linalg.inv(X)
print(Z_inverse)
print("Matriks Identitas")
I=X@X_inverse
print(I)

addition
[[ 2  4  6]
 [ 4  6 10]
 [12 14  9]]
substraction
[[ 0  0  0]
 [-4 -4 -2]
 [-2 -2 -9]]
Multiplication
[[ 1  4  9]
 [ 0  5 24]
 [35 48  0]]
dot product
[[30 36 42]
 [32 37 42]
 [29 40 51]]
dot product
[[30 36 42]
 [32 37 42]
 [29 40 51]]
Matriks Transpose
[[1 0 5]
 [2 1 6]
 [3 4 0]]
Matriks Inverse
[[-24.  18.   5.]
 [ 20. -15.  -4.]
 [ -5.   4.   1.]]
Matriks Identitas
[[ 1.00000000e+00 -2.66453526e-15  0.00000000e+00]
 [ 0.00000000e+00  1.00000000e+00  0.00000000e+00]
 [ 0.00000000e+00 -7.10542736e-15  1.00000000e+00]]


## **Bab 4: Indexing dan Slicing**

*4.1. Mengakses Elemen (Indexing)*

In [ ]:
import numpy as np

# Array 1 Dimensi
arr_1d = np.array([10, 11, 12, 13, 14])
print(f"Elemen ke-2 dari arr_1d: {arr_1d[1]}") # Mengambil elemen kedua

# Array 2 Dimensi (matriks)
arr_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
# Formatnya adalah array[baris, kolom]
print(f"Elemen pada baris 1, kolom 2 dari arr_2d: {arr_2d[0, 1]}") # Mengambil angka 2

Elemen ke-2 dari arr_1d: 11
Elemen pada baris 1, kolom 2 dari arr_2d: 2


*4.2. Mengambil Sebagian Elemen (Slicing)*

In [ ]:
import numpy as np

arr = np.arange(10) # Membuat array [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

# Mengambil elemen dari indeks 2 hingga sebelum indeks 5
# formatya array[start:end]
print(f"Slicing 1D: {arr[2:5]}")

# Mengambil sebagian dari array 2D
arr_2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

# Mengambil 2 baris pertama dan 2 kolom pertama
# Baris: 0 hingga sebelum 2. Kolom: 0 hingga sebelum 2.
print("Slicing 2D:\n", arr_2d[0:2, 0:2])

Slicing 1D: [2 3 4]
Slicing 2D:
 [[1 2]
 [4 5]]


*4.3. Boolean Indexing*

In [ ]:
import numpy as np

data_nilai = np.array([50, 85, 92, 65, 78, 95])

# Membuat kondisi boolean
lulus = data_nilai > 75
print(f"Kondisi boolean (lulus > 75): {lulus}")

# Menggunakan kondisi tersebut untuk menyaring array asli
print(f"Nilai yang lulus: {data_nilai[lulus]}")

Kondisi boolean (lulus > 75): [False  True  True False  True  True]
Nilai yang lulus: [85 92 78 95]


## **Bab 5: Studi Kasus: Contoh Latihan dengan Jawaban**

*Soal #1: Analisis Data Penjualan*

Buatlah sebuah array NumPy yang merepresentasikan data penjualan harian (dalam ribu Rupiah) sebuah toko selama seminggu: [150, 200, 180, 220, 250, 300, 280]. Hitunglah total penjualan selama seminggu, rata-rata penjualan harian, dan penjualan tertinggi yang terjadi.

In [ ]:
import numpy as np

# 1. Membuat array data penjualan
penjualan = np.array([150, 200, 180, 220, 250, 300, 280])

# 2. Menghitung statistik
total = np.sum(penjualan)
rata_rata = np.mean(penjualan)
tertinggi = np.max(penjualan)

print(f"Total Penjualan: Rp{total} ribu")
print(f"Rata-rata Penjualan: Rp{rata_rata:.2f} ribu")
print(f"Penjualan Tertinggi: Rp{tertinggi} ribu")

Total Penjualan: Rp1580 ribu
Rata-rata Penjualan: Rp225.71 ribu
Penjualan Tertinggi: Rp300 ribu


*Soal #2: Konversi Suhu Sensor*

Anda memiliki data suhu dari sebuah sensor dalam Celcius: [25.5, 26.1, -99.0, 27.2, 24.9, -99.0]. Nilai -99.0 adalah data error. Saring data untuk membuang nilai error, lalu konversi semua suhu yang valid ke Fahrenheit. Rumus: F=C
times
frac95+32.

In [ ]:
import numpy as np

# 1. Membuat array data suhu
suhu_celcius = np.array([25.5, 26.1, -99.0, 27.2, 24.9, -99.0])

# 2. Menyaring data error menggunakan boolean indexing
suhu_valid = suhu_celcius[suhu_celcius > 0]
print(f"Suhu valid (Celcius): {suhu_valid}")

# 3. Mengonversi suhu valid ke Fahrenheit (vectorization)
suhu_fahrenheit = suhu_valid * (9/5) + 32

print(f"Suhu dalam Fahrenheit: {np.round(suhu_fahrenheit, 2)}")

Suhu valid (Celcius): [25.5 26.1 27.2 24.9]
Suhu dalam Fahrenheit: [77.9  78.98 80.96 76.82]


**Bab 6: Latihan Mandiri**

*Studi Kasus: Diskon Harga*

Anda memiliki daftar harga beberapa barang dalam sebuah array NumPy: [50000, 120000, 75000, 250000, 95000]. Toko sedang mengadakan diskon sebesar 15%. Buatlah array baru yang berisi harga setelah diskon.

Tugas:

- Buat array NumPy untuk daftar harga tersebut.

- Hitung harga setelah diskon 15% untuk semua barang menggunakan operasi vectorization.

- Tampilkan array harga akhir.